File: docs/garden/src/smalltalk-messages/proxy.py
# Аналог doesNotUnderstand: в Python: __getattr__ вызывается,
# только когда обычный поиск атрибута не нашёл имя.


class LoggingProxy:
    def __init__(self, subject):
        # пишем в __dict__ напрямую, чтобы не зациклиться на собственных полях
        self.__dict__["_subject"] = subject
        self.__dict__["invocation_count"] = 0

    def __getattr__(self, name):          # «сообщение не понято»
        attr = getattr(self._subject, name)
        if not callable(attr):
            return attr

        def forward(*args):
            self.__dict__["invocation_count"] += 1
            print(f"performing {name}({', '.join(map(repr, args))})")
            return attr(*args)            # aMessage sendTo: subject
        return forward


p = LoggingProxy([])
p.append(3)
p.append(4)
print("count(3) =", p.count(3))
print("invocation_count =", p.invocation_count)   # найден обычным поиском

# Ограничение: неявные вызовы спецметодов ищутся в классе, а не через __getattr__
try:
    print(len(p))
except TypeError as e:
    print("len(p):", e)

File: docs/garden/src/smalltalk-messages/proxy.js
// Аналог doesNotUnderstand: в JavaScript: ловушка get объекта Proxy
// перехватывает КАЖДОЕ обращение к свойству, а не только неизвестные.
"use strict";

function loggingProxy(subject) {
  let invocationCount = 0;
  return new Proxy(subject, {
    get(target, name, receiver) {
      if (name === "invocationCount") return invocationCount;
      const value = Reflect.get(target, name, receiver);
      if (typeof value !== "function") return value;
      return (...args) => {
        invocationCount += 1;
        console.log(`performing ${String(name)}(${args.join(", ")})`);
        return value.apply(target, args);     // aMessage sendTo: subject
      };
    },
  });
}

const p = loggingProxy([]);
p.push(3);
p.push(4);
console.log("indexOf(4) =", p.indexOf(4));
console.log("length =", p.length);            // не функция — отдаём как есть
console.log("invocationCount =", p.invocationCount);
